ENH: Add Styler.pipe() method (#23229)

nmusolino opened the pull request:
Added `Styler.pipe()` method.  This allows users to easily apply and compose functions that operate on Styler objects, just like the `DataFrame.pipe()` method does for dataframes.

- [x] closes #23229
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

nmusolino committed:
Add Styler.pipe() method, akin to DataFrame.pipe()

nmusolino committed:
Add whatsnew entry for `Styler.pipe()`

pep8speaks commented:
Hello @nmusolino! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/io/formats/style.py`](https://github.com/nmusolino/pandas/blob/204705252397f2c7cd8ddb41ec8df61fa314de3c/pandas/io/formats/style.py) !

 - There are no PEP8 issues in the file [`pandas/tests/io/formats/test_style.py`](https://github.com/nmusolino/pandas/blob/204705252397f2c7cd8ddb41ec8df61fa314de3c/pandas/tests/io/formats/test_style.py) !

gfyoung commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -214,6 +214,7 @@ Other Enhancements
 - Compatibility with Matplotlib 3.0 (:issue:`22790`).
 - Added :meth:`Interval.overlaps`, :meth:`IntervalArray.overlaps`, and :meth:`IntervalIndex.overlaps` for determining overlaps between interval-like objects (:issue:`21998`)
 - :meth:`Timestamp.tz_localize`, :meth:`DatetimeIndex.tz_localize`, and :meth:`Series.tz_localize` have gained the ``nonexistent`` argument for alternative handling of nonexistent times. See :ref:`timeseries.timezone_nonexsistent` (:issue:`8917`)
+- :meth:`Styler.pipe` method added, to simplify application of user-defined functions that operate on stylers.
```
Reference the issue number.

gfyoung replied:
Also, you should create a mini-section explaining the enhancement in more detail so that end users can understand the benefit.

gfyoung commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,35 @@ class MyStyler(cls):
 
+    def pipe(self, func, *args, **kwargs):
+        """
+        Apply func(self, *args, **kwargs)
```
This should be a little more descriptive.  Apply `func` to what?  Also, for what purpose?

(yes, as reviewers, we can see what the rationale was, so just put that rationale into the docstring)

gfyoung commented on pandas/tests/io/formats/test_style.py:
```diff
@@ -1173,6 +1173,25 @@ def test_hide_columns_mult_levels(self):
+        def g(**kwargs):
+            assert 'styler' in kwargs
+            return kwargs['styler'].data
+        assert self.df.style.pipe((g, 'styler')) is self.df
```
Can you comment here explaining why you're using `is` instead of `assert_frame_equal` ?

gfyoung replied:
Also, add a comment explaining why you're creating this `g` function in the first place (perhaps a more descriptive function name is needed).

gfyoung commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,35 @@ class MyStyler(cls):
+        Styler.applymap
+        pandas.DataFrame.pipe
+        """
+        return com._pipe(self, func, *args, **kwargs)
```
Well, that was pretty easy. 🙂 

Makes me more in favor of this enhancement. 👍

nmusolino reviewed (commented):
Thank you, @gfyoung, for the prompt review and feedback.  All of your comments make sense, and I'll try to make the suggested changes.

nmusolino committed:
Simplify Styler.pipe() unit tests

nmusolino commented:
I have not pushed any changes yet, but this what the `pipe()` docs look like after local changes:

<img width="777" alt="pandas_styler_pipe_docs_1002" src="https://user-images.githubusercontent.com/3150861/47611100-4d030c00-da34-11e8-8a38-cdfd27babb89.png">

nmusolino committed:
Styler.pipe() docstring improvements

nmusolino committed:
Expand whatsnew entry per review feedback

nmusolino commented:
I'll summarize the changes made since the original push:

* Simplified the second part of the unit test
* Added "Notes" and "Examples" section to the docstring.  A rendering of the HTML documentation can be found in an earlier comment.
* Added a "version added" directive to the docstring (caught that one myself).
* Moved the "whatsnew" entry from a one-liner to a short paragraph/example.

I wasn't able to test how the whatsnew will be rendered.

At this point, the main questions are:
* Do the notes/examples for the new method look reasonable?  I'm open to any suggestions there.
* Should I leave this PR as five distinct commits, or manually rebase it into one commit?  I'm reasonably familiar with different approaches, but don't know what the common practice is for pandas.

datapythonista commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,76 @@ class MyStyler(cls):
 
+    def pipe(self, func, *args, **kwargs):
+        """
+        Apply func(self, \*args, \*\*kwargs), and return the result.
```
Can you add double backticks around `func(...)`

nmusolino replied:
Done.

datapythonista commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,76 @@ class MyStyler(cls):
+            ``callable`` that expects the Styler.
+        args : iterable, optional
+            positional arguments passed into ``func``.
+        kwargs : mapping, optional
```
Add the `*` before args and kwargs. I'd prefer `dict` intead of mapping. Parameter descriptions should start with a capital letter.

If you run `./scripts/validate_docstrings.py pandas.Styler.pipe` you should get a report with most of the docstring issues.

nmusolino replied:
Done.  As mentioned above, now validating.

datapythonista commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,76 @@ class MyStyler(cls):
+
+        Returns
+        -------
+        object : the value returned by ``func``.
```
Is this not always returning a `pandas.Styler` object?

The description should go in the next line (indented), and without the colon. And it should start by a capital letter.

nmusolino replied:
The return type is determined by `func`, the user-provided function.

datapythonista commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,76 @@ class MyStyler(cls):
+        --------
+        Styler.apply
+        Styler.applymap
+        pandas.DataFrame.pipe
```
`pandas.` prefix not needed. Please add a description on why those are relevant in the context of Styler.pipe.

nmusolino replied:
Done.

datapythonista commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,76 @@ class MyStyler(cls):
+        Styler.applymap
+        pandas.DataFrame.pipe
+
+        Notes
```
I think this should be the summary more than a Notes section.

nmusolino replied:
I'm inclined to leave this as "notes", for consistency with [pandas.DataFrame.pipe](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.pipe.html).  

Another reason to separate this from the function specification is that this section alludes to specific use cases for `pipe`, and conventions for the user-defined functions.  But I wouldn't want users to get the impression that they must use `pipe` like this -- there may be several other useful applications for the method that creative users will devise.

datapythonista replied:
We'll chage `DataFrame.pipe` then (you can open yourself another PR for it). We should show this in the examples (no code blocks should be in any other section). And the code needs to show real-world examples, and needs to be runnable.

nmusolino replied:
This was added at the request of @gfyoung [earlier in the review process](https://github.com/pandas-dev/pandas/pull/23384#discussion_r228727737), who asked:

> Also, for what purpose?
> 
> (yes, as reviewers, we can see what the rationale was from your PR, so just put that rationale into the docstring)

@TomAugspurger [suggested making this a code block](https://github.com/pandas-dev/pandas/pull/23384#issuecomment-435075033).

I think removing the code block may increase compliance with some developer rules for docstrings, but would make "rationale" part of the docstring less clear to end users.

@gfyoung, do you think the text-only rationale, without the code examples, is sufficient?

TomAugspurger replied:
I like the code block here.

datapythonista replied:
Not a big deal, but I don't think the code block will be run in the doctests, or validated for pep8 in the docstring validation. That's why I would prefer to have the code in the example, and not in code blocks (besides a personal preference on keeping all the code together).

TomAugspurger replied:
This code block is more for showing why `pipe` is still useful despite being functionally equivalent to function calls. It's not intended as an actual example.

datapythonista commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,76 @@ class MyStyler(cls):
+        ...     ''' Highlight the indicated element, and its containing row. '''
+        ...     cell, row = pd.IndexSlice[row_label, column], pd.IndexSlice[row_label, :]
+        ...     return (styler.set_properties(cell, **{'background-color': '#ffffcc'})
+        ...                   .set_properties(row, **{'background-color': '#ffccc0'}))
```
It'd be great if you can find a simpler example to illustrate `.pipe()`. Something like adding quotes around the values, or something as simple as that, so the function does not distract on what `.pipe()` does.

nmusolino replied:
That is a good suggestion.  I chose this to illustrate functionality that seems to belong in a function, and to use a function that could sensibly be applied twice.

BUT that's not the clearest example for the docs.  I'll use a much simpler function as you suggest.

datapythonista reviewed (commented):
Added some comments about the docstring.

nmusolino committed:
Revise whatsnew message

nmusolino committed:
Improve Styler.pipe() docstring

nmusolino commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -180,6 +180,26 @@ array, but rather an ``ExtensionArray``:
+       .pipe(set_summary_style)
+       .set_caption('Summary of results.'))
+
+Other classes that provide a ``pipe()`` method include DataFrame, Series, Groupby,
```
Done

nmusolino commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -180,6 +180,26 @@ array, but rather an ``ExtensionArray``:
 
+New ``Styler.pipe()`` method
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+A new method :meth:`pandas.formats.style.Styler.pipe()` was added (:issue:`#23229`).
```
Done

nmusolino commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -180,6 +180,26 @@ array, but rather an ``ExtensionArray``:
 
+New ``Styler.pipe()`` method
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+A new method :meth:`pandas.formats.style.Styler.pipe()` was added (:issue:`#23229`).
```
Done

nmusolino committed:
Final docstring tweak

nmusolino reviewed (commented):
Thanks for the feedback.  I've made changes as requested.  Here is an image of the updated documentation:

<img width="771" alt="screen shot 2018-10-29 at 10 31 36 pm" src="https://user-images.githubusercontent.com/3150861/47691960-82416280-dbca-11e8-8384-dbbbb8ade5c9.png">

The `pandas.DataFrame` links are not resolving, but I think that's because I am only building part of the docs, using `python make.py --single api`.

nmusolino commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,76 @@ class MyStyler(cls):
 
         return MyStyler
 
+    def pipe(self, func, *args, **kwargs):
```
I am traveling at the moment and am not able to copy-paste the output at this moment. But my recollection is:
1.  There were no docstring style issues identified by the `validate_docstrings.py` script, in my most recent version.
2.  The initial doctests on line 1260 (namely `>>> f(g(df.style.set_precision(3), arg1=a), arg2=b, arg3=c)`) and the subsequent rewritten version fail because the functions f and g are not defined.  I suppose the same thing happens in the analogous [DataFrame.pipe doctest](
https://github.com/pandas-dev/pandas/blob/0409521665bd436a10aea7e06336066bf07ff057/pandas/core/generic.py#L4230).
3. The later doctest on line 1287 executes as intended, but fails when the actual output (a string like `<pandas.io.formats.style.Styler at 0xaabcd538364>`) does not match the listed output in the doctest (no output included).  I could change this to make it formally pass, but it would require a `doctest: +ELLIPSIS` directive (to accommodate the memory address in the string), and it would be inconsistent with other doctests in this module (which don’t show the styler’s string representation as output).

nmusolino commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -180,6 +180,27 @@ array, but rather an ``ExtensionArray``:
 
+New ``Styler.pipe()`` method
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The dataframe Styler class has gained a :meth:`~pandas.io.formats.style.Styler.pipe` method
```
I can make those changes. I’ll also (a) add an anchor, like `.. _whatsnew_0240.enhancements.styler_pipe:` and (b) change the code block style to match everything else in the file.

TomAugspurger commented:
I'm fine with skipping the example with f and g. That could also be made
into a ..code-block:: python directive.

On Thu, Nov 1, 2018 at 8:03 AM Nicholas Musolino <notifications@github.com>
wrote:

> *@nmusolino* commented on this pull request.
> ------------------------------
>
> In pandas/io/formats/style.py
> <https://github.com/pandas-dev/pandas/pull/23384#discussion_r230063306>:
>
> > @@ -1222,6 +1222,76 @@ class MyStyler(cls):
>
>          return MyStyler
>
> +    def pipe(self, func, *args, **kwargs):
>
> I am traveling at the moment and am not able to copy-paste the output at
> this moment. But my recollection is:
>
>    1. There were no docstring style issues identified by the
>    validate_docstrings.py script, in my most recent version.
>    2. The initial doctests on line 1260 (namely >>>
>    f(g(df.style.set_precision(3), arg1=a), arg2=b, arg3=c)) and the
>    subsequent rewritten version fail because the functions f and g are not
>    defined. I suppose the same thing happens in the analogous DataFrame.pipe
>    doctest
>    <https://github.com/pandas-dev/pandas/blob/0409521665bd436a10aea7e06336066bf07ff057/pandas/core/generic.py#L4230>
>    .
>    3. The later doctest on line 1287 executes as intended, but fails when
>    the actual output (a string like <pandas.io.formats.style.Styler at
>    0xaabcd538364>) does not match the listed output in the doctest (no
>    output included). I could change this to make it formally pass, but it
>    would require a doctest: +ELLIPSIS directive (to accommodate the
>    memory address in the string), and it would be inconsistent with other
>    doctests in this module (which don’t show the styler’s string
>    representation as output).
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/23384#discussion_r230063306>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHImhjtI1ukK57kFlbGC6WyVYQWrNxks5uqw0lgaJpZM4X9o-F>
> .
>

TomAugspurger commented:
Merge conflict in 0.24.0, if you could update.

nmusolino commented:
I can resolve the merge conflict and turn the first example into a `code-block` instead of a doctest.  I’ll also clean up the whatsnew entry as discussed earlier.  I will push some changes Sunday night or Monday night.

nmusolino committed:
More whatsnew revisions

nmusolino committed:
Change non-runnable doctests to code blocks

nmusolino committed:
Merge branch 'master' into styler_pipe

nmusolino commented:
Here is the updated method documentation:
<img width="773" alt="screen shot 2018-11-05 at 9 42 02 pm" src="https://user-images.githubusercontent.com/3150861/48039650-bcb77c00-e143-11e8-840f-75127774662e.png">

Here is the output from the `validate_docstring.py` script:
```
$ python ./scripts/validate_docstrings.py pandas.io.formats.style.Styler.pipe

################################################################################
############### Docstring (pandas.io.formats.style.Styler.pipe)  ###############
################################################################################

Apply ``func(self, *args, **kwargs)``, and return the result.

.. versionadded:: 0.24.0

Parameters
----------
func : function
    Function to apply to the Styler.
    ``*args``, and ``**kwargs`` are passed into ``func``.
    Alternatively a ``(callable, data_keyword)`` tuple where
    ``data_keyword`` is a string indicating the keyword of
    ``callable`` that expects the Styler.
*args : iterable, optional
    Positional arguments passed into ``func``.
**kwargs : dict, optional
    Dictionary of keyword arguments passed into ``func``.

Returns
-------
result : object
    The value returned by ``func``.

See Also
--------
DataFrame.pipe : Analogous method for DataFrame.
Styler.apply : Apply a function row-wise, column-wise, or table-wise to
    modify the dataframe's styling.

Notes
-----
Like :meth:`DataFrame.pipe`, this method can simplify the
application of several user-defined functions to a styler.  Instead
of writing:

.. code-block:: python

    f(g(df.style.set_precision(3), arg1=a), arg2=b, arg3=c)

users can write:

.. code-block:: python

    (df.style.set_precision(3)
       .pipe(g, arg1=a)
       .pipe(f, arg2=b, arg3=c))

In particular, this allows users to define functions that take a
styler object, along with other parameters, and return the styler after
making styling changes (such as calling :meth:`Styler.apply` or
:meth:`Styler.set_properties`).  Using ``.pipe``, these user-defined
style "transformations" can be interleaved with calls to the built-in
Styler interface.

Examples
--------
>>> def set_standard_formatting(styler):
...     return (styler.set_properties(**{'text-align': 'right'})
...                   .format({'X': '{:.1%}'}))

The user-defined highlight function above can be called within a
sequence of other style modifications:

>>> df = pd.DataFrame({'A': list(range(-1, 4)), 'X': np.arange(0.2, 1.2, 0.2)})
>>> (df.style
...    .set_properties(subset=['X'], **{'background-color': 'yellow'})
...    .pipe(set_standard_formatting)
...    .set_caption("Results with column 'X' highlighted."))

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
	Examples do not pass tests

################################################################################
################################### Doctests ###################################
################################################################################

**********************************************************************
Line 65, in pandas.io.formats.style.Styler.pipe
Failed example:
    (df.style
       .set_properties(subset=['X'], **{'background-color': 'yellow'})
       .pipe(set_standard_formatting)
       .set_caption("Results with column 'X' highlighted."))
Expected nothing
Got:
    <pandas.io.formats.style.Styler object at 0x11c982b00>
```

datapythonista commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,78 @@ class MyStyler(cls):
+        Returns
+        -------
+        result : object
+            The value returned by ``func``.
```
Is this true? My understanding is that the result of `func` is used in the cells of the styler, and that a new styler is returned.

nmusolino replied:
I think you're confusing this method with some of the other Styler methods, like [Styler.applymap](https://pandas.pydata.org/pandas-docs/version/0.23.4/generated/pandas.io.formats.style.Styler.applymap.html).

datapythonista commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,78 @@ class MyStyler(cls):
 
+    def pipe(self, func, *args, **kwargs):
+        """
+        Apply ``func(self, *args, **kwargs)``, and return the result.
```
Not exactly true, what happens with the `Styler` instance where this is being called? Don't we modify its result?

nmusolino replied:
I'm not sure I understand your second question.

The user-supplied function `func` may or may not modify the styler; it may or may not have side-effects; and it may return the styler, some other result, or None.

Consider the `pandas.Series` case.  A user could write any of the following:

```
>>> s = pandas.Series(...)
>>> s.dropna().shift().pipe(normalize)        # Returns Series
>>> s.dropna().shift().pipe(trimmed_mean)     # Returns scalar
>>> s.dropna().shift().pipe(insert_initial_value)  # Unconventional, not "pand-onic"
>>> s.dropna().shift().pipe(save_to_disk)     # Returns string filename
```

datapythonista commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,78 @@ class MyStyler(cls):
+        The user-defined highlight function above can be called within a
+        sequence of other style modifications:
+
+        >>> df = pd.DataFrame({'A': list(range(-1, 4)), 'X': np.arange(0.2, 1.2, 0.2)})
```
Use something that looks real, no `range` fuctions, and no `a`, `x`, `foo` columns. We start to have many examples with animal names (e.g. `cat`, `penguin`, `falcon`...). I'd use that for consistency (hopefully at some point we have all the examples with similar data). Just have couple of columns that are percentages, and 2 or 3 rows.

Then rename `set_standard_formatting` to `percentage_format` (if I understand correctly, that's what it is, right?

nmusolino replied:
I made some changes here, and fixed the reference in the text, which was incorrect.  I changed the function name as you suggested, while trying to make it clear that the function was user-defined and application-specific.

I'll be honest, using animal names in examples seems a little silly to me, but that's up to the pandas developers.  

In this case, using numeric data rather than strings made more sense to me, since numeric data presents styling/formatting choices, and special values (min/max/null) can be highlighted.

datapythonista reviewed (changes requested):
Added some comments, still needs some work.

Make sure that the docstring validation does not have any error (and that the CI is green).

Thanks for working on this!

nmusolino commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,78 @@ class MyStyler(cls):
+            ``*args``, and ``**kwargs`` are passed into ``func``.
+            Alternatively a ``(callable, data_keyword)`` tuple where
+            ``data_keyword`` is a string indicating the keyword of
+            ``callable`` that expects the Styler.
```
The function supports `my_styler.pipe((f, 'styler')`, in which case, the implementation calls `f(styler=my_styler)` and returns the result.  I agree this is somewhat clumsy to document.

Yes, all the pipe methods share a common implementation, defined in `common.py`.

nmusolino commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,78 @@ class MyStyler(cls):
+
+        Returns
+        -------
+        result : object
```
As [mentioned previously](https://github.com/pandas-dev/pandas/pull/23384#discussion_r229156277), the return type is determined by func, the user-provided function.

nmusolino commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,78 @@ class MyStyler(cls):
+        *args : iterable, optional
+            Positional arguments passed into ``func``.
+        **kwargs : dict, optional
+            Dictionary of keyword arguments passed into ``func``.
```
Changed as suggested.  As a result of these changes, the `validate_docstring.py` script reports an error, but I agree this is better.

nmusolino commented on pandas/io/formats/style.py:
```diff
@@ -1222,6 +1222,78 @@ class MyStyler(cls):
+        ----------
+        func : function
+            Function to apply to the Styler.
+            ``*args``, and ``**kwargs`` are passed into ``func``.
```
Removed the line as suggested.

nmusolino committed:
Simplify docstring parameters/return sections

nmusolino committed:
Modify doctest example

nmusolino committed:
Merge branch 'styler_pipe' of github.com:nmusolino/pandas into styler_pipe

nmusolino committed:
Use natural subset for doctest example

nmusolino reviewed (commented):
I think we're reaching the point of diminishing marginal returns to the refinements in this PR, and it's about ready to move across the finish line.

Here's the latest docstring as rendered:

<img width="774" alt="screen shot 2018-11-10 at 11 54 21 am" src="https://user-images.githubusercontent.com/3150861/48303958-88dfac00-e4df-11e8-9e21-27efd0573e05.png">

TomAugspurger commented:
@nmusolino a few lines are too long: 
https://github.com/pandas-dev/pandas/pull/23384#issuecomment-433654308

nmusolino committed:
Ensure whatsnew entry runs cleanly

nmusolino committed:
Fix docstring line length, format issues

nmusolino commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -179,6 +179,31 @@ array, but rather an ``ExtensionArray``:
+        return (styler.format({'N': '{:,}', 'X': '{:.1%}'})
+                      .set_properties(**{'text-align': 'right'}))
+
+    (df.style.set_precision(1)
```
Done.

nmusolino commented:
@TomAugspurger, I've fixed the line length issues, and made sure the "whatsnew" documentation builds correctly, and looks reasonable.

Any other changes you would suggest?  And would you like me to squash, rebase, and force-push to clean up the commit history?

An updated docstring rendering can be found below.

<img width="761" alt="screen shot 2018-11-17 at 9 02 47 pm" src="https://user-images.githubusercontent.com/3150861/48667800-a1862e00-eaac-11e8-954a-1511400db6f0.png">

TomAugspurger reviewed (approved):
LGTM, if you could fix the conflict and ping when the CI passes I'll merge.

nmusolino committed:
Merge remote-tracking branch 'origin/master' into styler_pipe

nmusolino commented:
@TomAugspurger , I resolved the conflict, and CI is passing.  Would you be able to merge?

TomAugspurger commented:
Looks good, thanks @nmusolino!
